### PR TITLE
JDK-6251738: Want a top-level summary page that itemizes all spec documents referenced from javadocs (OEM spec)

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
This introduces support for a new `@spec` tag that can be used as either an inline tag or as a block tag. It is used to identify references to external specifications, in such a way that the references can be collected together on a new summary page, called "Other Specifications", available from either the static INDEX page or the interactive search box.

As an inline tag, the format is `{@spec url label}`, which is roughly translated to `<a href="url">label</a>` in the generated docs.

As a block tag, the format is simply

    @spec url label

which is handled in a manner analogous to

    @see <a href="url">label</a>

The tag is notable for being the first standard/supported tag that can be used as either an inline or block tag. (We have had support for bimodal non-standard/custom tags for a while, such as `{@jls}` and `{@jvms}`.) To support bimodal standard tags, some changes to `DocCommentParser` are incorporated here.

This change is only the _support_ for the new tag;  it does _not_ include any changes to API docs to _use_ the new tag.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Integration blocker
&nbsp;⚠️ The change requires a CSR request to be approved.

### Issues
 * [JDK-6251738](https://bugs.openjdk.java.net/browse/JDK-6251738): Want a top-level summary page that itemizes all spec documents referenced from javadocs (OEM spec)
 * [JDK-8226279](https://bugs.openjdk.java.net/browse/JDK-8226279): javadoc should support a new @spec tag

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to c777637f38c8e52bcfc7d27ac7e8bee83aee52d9
 * [Hannes Wallnöfer](https://openjdk.java.net/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to c777637f38c8e52bcfc7d27ac7e8bee83aee52d9

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/790/head:pull/790`
`$ git checkout pull/790`

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build changes look good.

**Review by hns — APPROVED** [review]
It's very instructive to see a new tag introduced from scratch with all the accompanying infrastructure and tests. The changes looks good to me, but I need to stress than I'm not an expert on some of the affected areas such as the jdk.compiler parts.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java [hunk 1]
```diff
@@ -302,6 +303,17 @@ DocCommentTree newDocCommentTree(List<? extends DocTree> fullBody,
      */
     SinceTree newSinceTree(List<? extends DocTree> text);
 
+    /**
+     * Creates a new {@code SpecTree} object, to represent a {@code @spec} tag.
+     *
+     * @param inline whether this is instance is an inline tag
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java:309]: Should be: 
`* @param inline whether this instance is an inline tag`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletManager.java [hunk 2]
```diff
@@ -733,7 +734,7 @@ private void showTaglets(PrintStream out) {
         taglets.addAll(allTaglets.values());
 
         for (Taglet t : taglets) {
-            String name = t.isInlineTag() ? "{@" + t.getName() + "}" : "@" + t.getName();
+            String name = t.isBlockTag() ? "@" + t.getName() : "{@" + t.getName() + "}";
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletManager.java:737]: Out of curiosity, why did you flip that?

**Review by pavelrappo — CHANGES_REQUESTED** [review]
1. Thanks for incorporating my previous offline feedback.
2. Since Hannes and Erik seem to have looked at everything else, I looked mostly at changes in `src/jdk.compiler/share/classes/com/sun/source/**`, which are good!
3. There should be a corresponding but separate change to "Documentation Comment Specification for the Standard Doclet".
4. Can we use this new `@since` tag to refer to the spec at `com/sun/tools/javac/parser/DocCommentParser.java:1116`?
5. Should we specify in `com.sun.source.doctree.SpecTree` that both `url` and `label` parts are mandatory?
6. `DCSpec extends DCEndPosTree`, sigh... Although that is not a public API, this design suggests we could improve that abstraction sometime later.

**Comment by jonathan-gibbons** [comment]
Closing pull request, until we better decide the contents of the spec page

**Pull request closed without merging**